File: labs/mathlogic/lab2/lab2_3.py
from typing import TypeVar, Generic, Optional, List
from lab2_1 import StructureNode

T = TypeVar('T')

class SinglyLinkedList(Generic[T]):
    def __init__(self):
        self.head: Optional[StructureNode[T]] = None
        self._size: int = 0

    # можно и без этого, но мне удобнее так
    def _is_empty(self):
        return self._size == 0

    def prepend(self, data: T):
        next_node = self.head
        self.head = StructureNode(data)
        self.head.next = next_node
        self._size = self._size + 1

    def append(self, data: T):
        current: Optional[StructureNode[T]] = self.head
        if self._is_empty():
            self.prepend(data)
            return
        
        while current.next is not None:
            current = current.next
        
        current.next = StructureNode(data)
        self._size = self._size + 1

    # я не знаю как это будет работать с объектами, так как они сравниваются по ссылкам
    # предположим, что мы пока что работаем только с примитивами
    def find(self, data: T):
        if self._is_empty():
            raise Exception("List is empty!")

        index = 0
        node: Optional[StructureNode[T]] = self.head

        while node is not None:
            if node.data == data:
                return index
            node = node.next
            index += 1

        return -1

    def insert(self, data: T, index: int):
        if index <= 0 or self._is_empty():
            self.prepend(data)
            return
        if index >= self._size:
            self.append(data)
            return
        
        counter = 0
        node: Optional[StructureNode[T]] = self.head

        while counter != index - 1:
            node = node.next
            counter += 1

        next_node = node.next
        node.next = StructureNode(data)
        node.next.next = next_node
        self._size = self._size + 1

    # не очень понятно, тут удалять только 1 или все случаи?
    def remove(self, data: T):
        if self._is_empty():
            return

        node: Optional[StructureNode[T]] = None
        next_node: Optional[StructureNode[T]] = self.head

        while next_node is not None:
            if next_node.data == data:
                next_ref = next_node.next
                self._size = self._size - 1
                if node is None:
                    self.head = next_ref
                    return
                node.next = next_ref
                return
            node = next_node
            next_node = next_node.next

    # не очень понимаю, тут они выводятся по одному или всем скопом?
    # return для явного вывода, чтобы я мог это в верстке показать
    def display(self):
        if self._is_empty():
            print("")
            return "null"

        node: Optional[StructureNode[T]] = self.head
        nodes: List[T] = []

        while node is not None:
            nodes.append(node.data)
            node = node.next

        output = " -> ".join(map(str, nodes)) + " -> null"
        print(output)
        return output

File: labs/mathlogic/lab2/lab2_3.ts
import { StructureNode } from "./lab2_1";

//tail намеренно не использую
export class SinglyLinkedList<T> {
  head: StructureNode<T> | null = null;
  private _size: number = 0;

  //можно и без этого, но мне удобнее так
  private _is_empty() {
    return this._size == 0;
  }

  prepend(data: T) {
    const next = this.head;
    this.head = new StructureNode(data);
    this.head.next = next;
    this._size = this._size + 1;
  }

  append(data: T) {
    let current: StructureNode<T> | null = this.head;
    if (this._is_empty()) {
      this.prepend(data);
      return;
    }
    while (current!.next != null) {
      current = current!.next;
    }
    current!.next = new StructureNode(data);
    this._size = this._size + 1;
  }

  //я не знаю как это будет работать с объектами, так как они сравниваются по ссылкам
  //предположим, что мы пока что работаем только с примитивами
  find(data: T) {
    if (this._is_empty()) throw new Error("List is empty!");

    let index = 0;
    let node: StructureNode<T> | null = this.head!;

    while (!!node) {
      if (node.data == data) {
        return index;
      }
      node = node.next;
      index++;
    }

    return -1;
  }

  insert(data: T, index: number) {
    if (index <= 0 || this._is_empty()) {
      this.prepend(data);
      return;
    }
    if (index >= this._size) {
      this.append(data);
      return;
    }
    let counter = 0;
    let node: StructureNode<T> | null = this.head!;

    while (counter != index - 1) {
      node = node!.next;
      counter++;
    }

    const next = node!.next;
    node!.next = new StructureNode(data);
    node!.next.next = next;
    this._size = this._size + 1;
  }

  //не очень понятно, тут удалять только 1 или все случаи?
  remove(data: T) {
    if (this._is_empty()) return;

    let node: StructureNode<T> | null = null;
    let nextNode: StructureNode<T> | null = this.head!;

    while (!!nextNode) {
      if (nextNode.data == data) {
        const next = nextNode.next;
        this._size = this._size - 1;
        if (!node) {
          this.head = next;
          return;
        }
        node.next = next;
        return;
      }
      node = nextNode;
      nextNode = nextNode.next;
    }
  }

  //не очень понимаю, тут они выводятся по одному или всем скопом?
  //return для явного вывода, чтобы я мог это в верстке показать
  display() {
    if (this._is_empty()) {
      console.log("");
      return "null";
    }

    let node: StructureNode<T> | null = this.head;
    let nodes: T[] = [];

    while (!!node) {
      nodes.push(node.data);
      node = node.next;
    }

    const output = nodes.join(" -> ") + " -> null";
    console.log(output);
    return output;
  }
}
